File: web/frontend/assets/js/secondary-tool-contracts.js
/** Strict client-side contracts for Stress, Diagnose, Apply and Struct Lint. */
(function (root) {
  'use strict';

  var CONTRACT_VERSION = 'secondary-tools-v2';
  var REQUEST_ID = /^[A-Za-z0-9][A-Za-z0-9_-]{11,63}$/;
  var CANDIDATE_ID = /^[A-Za-z0-9][A-Za-z0-9._:-]{0,127}$/;
  var CLAIM_ID = /^lint-[0-9a-f]{16}$/;

  function plain(value) {
    if (!value || Object.prototype.toString.call(value) !== '[object Object]') return false;
    var proto = Object.getPrototypeOf(value);
    return proto === Object.prototype || proto === null;
  }

  function exact(value, keys) {
    if (!plain(value)) return false;
    var actual = Object.keys(value).sort();
    var expected = keys.slice().sort();
    return actual.length === expected.length && actual.every(function (key, i) {
      return key === expected[i];
    });
  }

  function text(value, min, max) {
    return typeof value === 'string' && value.trim().length >= min && value.length <= max;
  }

  function nullableText(value, max) {
    return value === null || text(value, 1, max);
  }

  function normalizeSource(value) {
    return String(value || '').normalize('NFKC').replace(/\s+/g, ' ').trim();
  }

  function createRequestId(prefix) {
    var head = String(prefix || 'secondary').replace(/[^A-Za-z0-9_-]/g, '').slice(0, 16) || 'secondary';
    var random = '';
    try {
      if (root.crypto && typeof root.crypto.randomUUID === 'function') {
        random = root.crypto.randomUUID().replace(/-/g, '');
      }
    } catch (error) { random = ''; }
    if (!random) {
      random = Date.now().toString(36) + Math.random().toString(36).slice(2, 14);
    }
    return (head + '-' + random).slice(0, 64);
  }

  function validateEvidence(value, expectedLabel, sourceKind) {
    if (!exact(value, [
      'schema_version', 'evidence_level', 'candidate', 'source', 'result',
      'independence', 'counterexamples', 'ledger'
    ])) return false;
    if (value.schema_version !== 'evidence-envelope-v1' || value.evidence_level !== 'candidate') return false;
    if (!exact(value.candidate, ['status', 'kind', 'label', 'score'])) return false;
    if (value.candidate.status !== 'recorded' || !text(value.candidate.kind, 1, 100)) return false;
    if (expectedLabel !== undefined && value.candidate.label !== expectedLabel) return false;
    if (!nullableText(value.candidate.label, 1000) || value.candidate.score !== null) return false;

    if (!exact(value.source, ['status', 'kind', 'label', 'url', 'source_review'])) return false;
    if (['recorded', 'not_recorded'].indexOf(value.source.status) === -1) return false;
    if (['internal_kb', 'not_recorded'].indexOf(value.source.kind) === -1) return false;
    if (sourceKind && value.source.kind !== sourceKind) return false;
    if (!nullableText(value.source.label, 1000) || value.source.url !== null || value.source.source_review !== null) return false;

    if (!exact(value.result, ['status', 'provenance', 'verdict', 'summary'])) return false;
    if (['recorded', 'not_recorded'].indexOf(value.result.status) === -1) return false;
    if (['NOT_TESTED', 'INTERNAL_AI_SCREEN'].indexOf(value.result.provenance) === -1) return false;
    if (['NOT_TESTED', 'INCONCLUSIVE'].indexOf(value.result.verdict) === -1) return false;
    if (!nullableText(value.result.summary, 1000)) return false;
    if (sourceKind === 'internal_kb') {
      if (value.source.status !== 'recorded' || value.source.label !== 'Structural KB record' ||
          value.result.status !== 'not_recorded' || value.result.provenance !== 'NOT_TESTED' ||
          value.result.verdict !== 'NOT_TESTED') return false;
    } else if (value.source.status !== 'not_recorded' || value.source.kind !== 'not_recorded' ||
               value.source.label !== null || value.result.status !== 'recorded' ||
               value.result.provenance !== 'INTERNAL_AI_SCREEN' ||
               value.result.verdict !== 'INCONCLUSIVE') return false;

    if (!exact(value.independence, ['status', 'kind', 'summary'])) return false;
    if (value.independence.status !== 'not_recorded' || value.independence.kind !== 'not_recorded') return false;
    if (!nullableText(value.independence.summary, 1000)) return false;

    if (!exact(value.counterexamples, ['status', 'summary'])) return false;
    if (['not_recorded', 'gap_recorded'].indexOf(value.counterexamples.status) === -1) return false;
    if (!nullableText(value.counterexamples.summary, 1000)) return false;

    if (!exact(value.ledger, ['status', 'claim_id', 'version', 'recorded_at', 'artifact_sha256', 'url'])) return false;
    if (value.ledger.status !== 'not_recorded') return false;
    return ['claim_id', 'version', 'recorded_at', 'artifact_sha256', 'url'].every(function (key) {
      return value.ledger[key] === null;
    });
  }

  function validateCandidateReference(value) {
    if (value === null) return true;
    if (!exact(value, [
      'id', 'name', 'domain', 'description', 'retrieval_rank', 'candidate_note', 'evidence'
    ])) return false;
    return CANDIDATE_ID.test(value.id) && text(value.name, 1, 200) &&
      typeof value.domain === 'string' && value.domain.length <= 120 &&
      typeof value.description === 'string' && value.description.length <= 600 &&
      Number.isInteger(value.retrieval_rank) && value.retrieval_rank >= 1 && value.retrieval_rank <= 30 &&
      nullableText(value.candidate_note, 600) &&
      validateEvidence(value.evidence, value.name, 'internal_kb');
  }

  function validateStressPayload(value, requestId, claim) {
    if (!exact(value, [
      'contract_version', 'request_id', 'claim', 'screening_outcome', 'screening_basis',
      'source', 'target', 'structural_correspondences', 'weakest_link', 'rationale',
      'candidate_reference', 'evidence'
    ])) return null;
    if (value.contract_version !== CONTRACT_VERSION || value.request_id !== requestId || value.claim !== claim) return null;
    if (['not_broken_in_screen', 'breaks_in_screen', 'condition_dependent'].indexOf(value.screening_outcome) === -1) return null;
    if (value.screening_basis !== 'internal_ai_red_team' || !text(value.source, 1, 400) || !text(value.target, 1, 400)) return null;
    if (!Array.isArray(value.structural_correspondences) || value.structural_correspondences.length < 1 || value.structural_correspondences.length > 12) return null;
    var validCorrespondences = value.structural_correspondences.every(function (item) {
      return exact(item, ['claim', 'screening_outcome', 'stress_result']) &&
        text(item.claim, 1, 600) && text(item.stress_result, 1, 1000) &&
        ['not_broken', 'breaks', 'uncertain'].indexOf(item.screening_outcome) !== -1;
    });
    if (!validCorrespondences || !text(value.weakest_link, 1, 1000) || !text(value.rationale, 1, 1200)) return null;
    if (!validateCandidateReference(value.candidate_reference)) return null;
    return validateEvidence(value.evidence, claim) ? value : null;
  }

  function validateState(value) {
    return exact(value, ['state_id', 'name', 'definition', 'typical_signal']) &&
      /^[a-z][a-z0-9_]{2,63}$/.test(value.state_id) && text(value.name, 1, 120) &&
      text(value.definition, 1, 500) && text(value.typical_signal, 1, 500);
  }

  function validTextList(value, maxItems, maxChars) {
    return Array.isArray(value) && value.length <= maxItems && value.every(function (item) {
      return text(item, 1, maxChars);
    });
  }

  function validateDiagnosePayload(value, requestId, situation) {
    if (!exact(value, [
      'contract_version', 'request_id', 'situation', 'assessment_kind', 'primary_state',
      'secondary_state', 'reasoning', 'evolution', 'signals_to_watch', 'recommendations',
      'candidate_reference', 'evidence'
    ])) return null;
    if (value.contract_version !== CONTRACT_VERSION || value.request_id !== requestId || value.situation !== situation) return null;
    if (value.assessment_kind !== 'structural_state_hypothesis' || !validateState(value.primary_state)) return null;
    if (value.secondary_state !== null && !validateState(value.secondary_state)) return null;
    if (value.secondary_state && value.secondary_state.state_id === value.primary_state.state_id) return null;
    if (!text(value.reasoning, 1, 1500) || !text(value.evolution, 1, 1200)) return null;
    if (!validTextList(value.signals_to_watch, 6, 500) || !validTextList(value.recommendations, 5, 800)) return null;
    if (!validateCandidateReference(value.candidate_reference)) return null;
    return validateEvidence(value.evidence, situation) ? value : null;
  }

  function validateMethodCandidate(value, expectedRank) {
    if (!exact(value, [
      'id', 'name', 'domain', 'type_id', 'description', 'retrieval_rank',
      'candidate_note', 'evidence'
    ])) return false;
    return CANDIDATE_ID.test(value.id) && text(value.name, 1, 200) &&
      typeof value.domain === 'string' && value.domain.length <= 120 &&
      typeof value.type_id === 'string' && value.type_id.length <= 120 &&
      typeof value.description === 'string' && value.description.length <= 600 &&
      value.retrieval_rank === expectedRank && nullableText(value.candidate_note, 240) &&
      validateEvidence(value.evidence, value.name, 'internal_kb');
  }

  function validateApplyPayload(value, requestId, method) {
    if (!exact(value, [
      'contract_version', 'request_id', 'method', 'signature', 'signature_origin',
      'keywords', 'count', 'candidates', 'evidence'
    ])) return null;
    if (value.contract_version !== CONTRACT_VERSION || value.request_id !== requestId || value.method !== method) return null;
    if (!text(value.signature, 1, 600) || ['model_generated', 'input_fallback'].indexOf(value.signature_origin) === -1) return null;
    if (!validTextList(value.keywords, 6, 30) || !Number.isInteger(value.count) || value.count < 0 || value.count > 20) return null;
    if (!Array.isArray(value.candidates) || value.count !== value.candidates.length) return null;
    var ids = Object.create(null);
    for (var i = 0; i < value.candidates.length; i += 1) {
      var candidate = value.candidates[i];
      if (!validateMethodCandidate(candidate, i + 1) || ids[candidate.id]) return null;
      ids[candidate.id] = true;
    }
    return validateEvidence(value.evidence, method) ? value : null;
  }

  function validateLintPayload(value, requestId, documentText) {
    if (!exact(value, [
      'contract_version', 'request_id', 'screening_kind', 'summary', 'claims', 'evidence'
    ])) return null;
    if (value.contract_version !== CONTRACT_VERSION || value.request_id !== requestId ||
        value.screening_kind !== 'internal_ai_document_screen' || !text(value.summary, 1, 1200)) return null;
    if (!Array.isArray(value.claims) || value.claims.length > 30) return null;
    var normalizedDocument = normalizeSource(documentText);
    var ids = Object.create(null);
    for (var i = 0; i < value.claims.length; i += 1) {
      var claim = value.claims[i];
      if (!exact(claim, [
        'claim_id', 'quote', 'claim_type', 'structure', 'failure_mode', 'review_priority',
        'suggestion', 'reference_candidate', 'evidence'
      ])) return null;
      if (!CLAIM_ID.test(claim.claim_id) || ids[claim.claim_id] ||
          !text(claim.quote, 1, 600) || normalizedDocument.indexOf(normalizeSource(claim.quote)) === -1) return null;
      ids[claim.claim_id] = true;
      if (['assumption', 'analogy', 'causal_judgment'].indexOf(claim.claim_type) === -1 ||
          ['high', 'medium', 'low'].indexOf(claim.review_priority) === -1) return null;
      if (!text(claim.structure, 1, 800) || !text(claim.failure_mode, 1, 800) || !text(claim.suggestion, 1, 800)) return null;
      if (!validateCandidateReference(claim.reference_candidate) || !validateEvidence(claim.evidence, claim.quote)) return null;
    }
    return validateEvidence(value.evidence, '用户提交的策略文档') ? value : null;
  }

  var api = {
    CONTRACT_VERSION: CONTRACT_VERSION,
    createRequestId: createRequestId,
    validateEvidence: validateEvidence,
    validateStressPayload: validateStressPayload,
    validateDiagnosePayload: validateDiagnosePayload,
    validateApplyPayload: validateApplyPayload,
    validateLintPayload: validateLintPayload
  };

  if (typeof module !== 'undefined' && module.exports) module.exports = api;
  root.SecondaryToolContracts = api;
}(typeof window !== 'undefined' ? window : globalThis));

File: web/backend/services/secondary_tool_contracts.py
"""Strict public contracts for the four secondary research tools.

These surfaces are useful hypothesis generators, not experimental validators.
The contract therefore keeps every result at the ``candidate`` evidence level,
binds every response to one client request id, and never exposes retrieval
scores as probabilities or model self-confidence as calibration.
"""
from __future__ import annotations

import re
import uuid
from typing import Any, Literal, Mapping, Optional

from pydantic import BaseModel, ConfigDict, Field, model_validator

from .evidence_envelope import (
    ResultProvenance,
    build_evidence_envelope,
)
from .scope_guard import is_out_of_scope


CONTRACT_VERSION = "secondary-tools-v2"
REQUEST_ID_PATTERN = re.compile(r"^[A-Za-z0-9][A-Za-z0-9_-]{11,63}$")
_STRUCTURAL_CONTEXT = re.compile(
    r"(?:公司|团队|组织|项目|策略|方案|业务|用户|系统|方法|算法|模型|机制|反馈|"
    r"流程|决策|结构|增长|市场|产品|网络|供应链|实验|数据|干预|迁移|类比|"
    r"company|team|organisation|organization|project|strategy|business|user|"
    r"system|method|algorithm|model|mechanism|feedback|process|decision|growth)",
    re.IGNORECASE,
)
_CLAUSE_BREAK = re.compile(r"[,，。;；!?！？\n]")
_DIRECT_FACT_QUERY = re.compile(
    r"(?:是什么|是哪里|在哪里|是谁|多少|几岁|几点|何时|什么时候|"
    r"首都|天气|人口|面积|定义)(?:[^。！？!?]{0,30})(?:[。！？!?]|$)",
    re.IGNORECASE,
)


def ensure_request_id(value: Optional[str]) -> str:
    """Return a validated client id or a server-generated compatibility id."""
    if value is None:
        return uuid.uuid4().hex
    if not isinstance(value, str) or not REQUEST_ID_PATTERN.fullmatch(value):
        raise ValueError("client_request_id has an invalid shape")
    return value


def secondary_scope_guard(value: str) -> tuple[bool, str]:
    """Close trivial-query suffix bypasses without changing primary Search."""
    out_of_scope, reason = is_out_of_scope(value)
    if out_of_scope:
        return out_of_scope, reason
    has_context = bool(_STRUCTURAL_CONTEXT.search(value))
    first_clause = _CLAUSE_BREAK.split(value, maxsplit=1)[0].strip()
    if first_clause and first_clause != value and not has_context:
        clause_out, clause_reason = is_out_of_scope(first_clause)
        if clause_out:
            return True, clause_reason
    if not has_context and _DIRECT_FACT_QUERY.search(value):
        return True, "trivia"
    return False, "ok"


class _StrictModel(BaseModel):
    model_config = ConfigDict(extra="forbid", strict=True, str_strip_whitespace=True)


class EvidenceCandidate(_StrictModel):
    status: Literal["recorded"]
    kind: str = Field(min_length=1, max_length=100)
    label: Optional[str] = Field(default=None, max_length=1_000)
    # Numeric retrieval scores are not calibrated transfer probabilities.
    # Keep them out of every secondary-tool response, including nested fields.
    score: None = None


class EvidenceSource(_StrictModel):
    status: Literal["recorded", "not_recorded"]
    kind: Literal["not_recorded", "internal_kb"]
    label: Optional[str] = Field(default=None, max_length=1_000)
    url: None = None
    source_review: None = None


class EvidenceResult(_StrictModel):
    status: Literal["recorded", "not_recorded"]
    provenance: Literal["NOT_TESTED", "INTERNAL_AI_SCREEN"]
    verdict: Literal["NOT_TESTED", "INCONCLUSIVE"]
    summary: Optional[str] = Field(default=None, max_length=1_000)


class EvidenceIndependence(_StrictModel):
    status: Literal["not_recorded"]
    kind: Literal["not_recorded"]
    summary: Optional[str] = Field(default=None, max_length=1_000)


class EvidenceCounterexamples(_StrictModel):
    status: Literal["not_recorded", "gap_recorded"]
    summary: Optional[str] = Field(default=None, max_length=1_000)


class EvidenceLedger(_StrictModel):
    status: Literal["not_recorded"]
    claim_id: Optional[str] = Field(default=None, max_length=200)
    version: Optional[str] = Field(default=None, max_length=100)
    recorded_at: Optional[str] = Field(default=None, max_length=100)
    artifact_sha256: Optional[str] = Field(default=None, max_length=64)
    url: Optional[str] = Field(default=None, max_length=2_048)


class CandidateEvidenceEnvelope(_StrictModel):
    schema_version: Literal["evidence-envelope-v1"]
    evidence_level: Literal["candidate"]
    candidate: EvidenceCandidate
    source: EvidenceSource
    result: EvidenceResult
    independence: EvidenceIndependence
    counterexamples: EvidenceCounterexamples
    ledger: EvidenceLedger

    @model_validator(mode="after")
    def coherent_candidate_state(self) -> "CandidateEvidenceEnvelope":
        if self.source.kind == "internal_kb":
            if (
                self.source.status != "recorded"
                or self.source.label != "Structural KB record"
                or self.result.status != "not_recorded"
                or self.result.provenance != "NOT_TESTED"
                or self.result.verdict != "NOT_TESTED"
            ):
                raise ValueError("KB candidate evidence state is inconsistent")
        elif (
            self.source.status != "not_recorded"
            or self.source.label is not None
            or self.result.status != "recorded"
            or self.result.provenance != "INTERNAL_AI_SCREEN"
            or self.result.verdict != "INCONCLUSIVE"
        ):
            raise ValueError("internal screen evidence state is inconsistent")
        if any((
            self.ledger.claim_id,
            self.ledger.version,
            self.ledger.recorded_at,
            self.ledger.artifact_sha256,
            self.ledger.url,
        )):
            raise ValueError("candidate evidence cannot carry an unbound ledger")
        return self


def internal_screen_evidence(*, kind: str, label: str) -> dict[str, Any]:
    """Candidate envelope for an LLM screen with no empirical validation."""
    raw = build_evidence_envelope(
        candidate_kind=kind,
        candidate_label=label,
        result_provenance=ResultProvenance.INTERNAL_AI_SCREEN,
        result_verdict="INCONCLUSIVE",
        result_summary="内部模型筛查；尚未经过独立数据或实验验证。",
        counterexample_status="gap_recorded",
        counterexample_summary="需要用反例搜索或现实数据继续检验。",
    )
    return CandidateEvidenceEnvelope.model_validate(raw).model_dump()


def kb_candidate_evidence(
    item: Mapping[str, Any], *, counterexample: Optional[str] = None
) -> dict[str, Any]:
    """Candidate envelope for one allowlisted internal-KB retrieval row."""
    raw = build_evidence_envelope(
        candidate_kind="retrieval_candidate",
        candidate_label=item.get("name") if isinstance(item.get("name"), str) else None,
        source_kind="internal_kb",
        source_label="Structural KB record",
        counterexample_status="gap_recorded" if counterexample else "not_recorded",
        counterexample_summary=counterexample,
    )
    return CandidateEvidenceEnvelope.model_validate(raw).model_dump()


class CandidateReference(_StrictModel):
    id: str = Field(pattern=r"^[A-Za-z0-9][A-Za-z0-9._:-]{0,127}$")
    name: str = Field(min_length=1, max_length=200)
    domain: str = Field(max_length=120)
    description: str = Field(max_length=600)
    retrieval_rank: int = Field(ge=1, le=30)
    candidate_note: Optional[str] = Field(default=None, max_length=600)
    evidence: CandidateEvidenceEnvelope

    @model_validator(mode="after")
    def bind_evidence(self) -> "CandidateReference":
        if self.evidence.candidate.label != self.name:
            raise ValueError("candidate evidence label is not bound to the record")
        if self.evidence.source.kind != "internal_kb":
            raise ValueError("candidate reference must come from the internal KB")
        return self


class StressCorrespondence(_StrictModel):
    claim: str = Field(min_length=1, max_length=600)
    screening_outcome: Literal["not_broken", "breaks", "uncertain"]
    stress_result: str = Field(min_length=1, max_length=1_000)


class StressTestResponse(_StrictModel):
    contract_version: Literal["secondary-tools-v2"]
    request_id: str = Field(pattern=r"^[A-Za-z0-9][A-Za-z0-9_-]{11,63}$")
    claim: str = Field(min_length=4, max_length=600)
    screening_outcome: Literal[
        "not_broken_in_screen", "breaks_in_screen", "condition_dependent"
    ]
    screening_basis: Literal["internal_ai_red_team"]
    source: str = Field(min_length=1, max_length=400)
    target: str = Field(min_length=1, max_length=400)
    structural_correspondences: list[StressCorrespondence] = Field(
        min_length=1, max_length=12
    )
    weakest_link: str = Field(min_length=1, max_length=1_000)
    rationale: str = Field(min_length=1, max_length=1_200)
    candidate_reference: Optional[CandidateReference]
    evidence: CandidateEvidenceEnvelope

    @model_validator(mode="after")
    def bind_screen(self) -> "StressTestResponse":
        if self.evidence.candidate.label != self.claim:
            raise ValueError("stress evidence is not bound to the request")
        return self


class StructuralState(_StrictModel):
    state_id: str = Field(pattern=r"^[a-z][a-z0-9_]{2,63}$")
    name: str = Field(min_length=1, max_length=120)
    definition: str = Field(min_length=1, max_length=500)
    typical_signal: str = Field(min_length=1, max_length=500)


class DiagnoseResponse(_StrictModel):
    contract_version: Literal["secondary-tools-v2"]
    request_id: str = Field(pattern=r"^[A-Za-z0-9][A-Za-z0-9_-]{11,63}$")
    situation: str = Field(min_length=12, max_length=1_500)
    assessment_kind: Literal["structural_state_hypothesis"]
    primary_state: StructuralState
    secondary_state: Optional[StructuralState]
    reasoning: str = Field(min_length=1, max_length=1_500)
    evolution: str = Field(min_length=1, max_length=1_200)
    signals_to_watch: list[str] = Field(max_length=6)
    recommendations: list[str] = Field(max_length=5)
    candidate_reference: Optional[CandidateReference]
    evidence: CandidateEvidenceEnvelope

    @model_validator(mode="after")
    def distinct_states(self) -> "DiagnoseResponse":
        if self.secondary_state and self.secondary_state.state_id == self.primary_state.state_id:
            raise ValueError("primary and secondary state must differ")
        if self.evidence.candidate.label != self.situation:
            raise ValueError("diagnosis evidence is not bound to the request")
        return self


class MethodCandidate(_StrictModel):
    id: str = Field(pattern=r"^[A-Za-z0-9][A-Za-z0-9._:-]{0,127}$")
    name: str = Field(min_length=1, max_length=200)
    domain: str = Field(max_length=120)
    type_id: str = Field(max_length=120)
    description: str = Field(max_length=600)
    retrieval_rank: int = Field(ge=1, le=20)
    candidate_note: Optional[str] = Field(default=None, max_length=240)
    evidence: CandidateEvidenceEnvelope

    @model_validator(mode="after")
    def bind_evidence(self) -> "MethodCandidate":
        if self.evidence.candidate.label != self.name:
            raise ValueError("method candidate evidence label is not bound")
        if self.evidence.source.kind != "internal_kb":
            raise ValueError("method candidate must be an internal KB row")
        return self


class MethodApplyResponse(_StrictModel):
    contract_version: Literal["secondary-tools-v2"]
    request_id: str = Field(pattern=r"^[A-Za-z0-9][A-Za-z0-9_-]{11,63}$")
    method: str = Field(min_length=4, max_length=1_000)
    signature: str = Field(min_length=1, max_length=600)
    signature_origin: Literal["model_generated", "input_fallback"]
    keywords: list[str] = Field(max_length=6)
    count: int = Field(ge=0, le=20)
    candidates: list[MethodCandidate] = Field(max_length=20)
    evidence: CandidateEvidenceEnvelope

    @model_validator(mode="after")
    def bind_candidate_set(self) -> "MethodApplyResponse":
        if self.count != len(self.candidates):
            raise ValueError("candidate count mismatch")
        ids = [candidate.id for candidate in self.candidates]
        ranks = [candidate.retrieval_rank for candidate in self.candidates]
        if len(ids) != len(set(ids)) or ranks != list(range(1, len(ranks) + 1)):
            raise ValueError("candidate identity or retrieval rank is invalid")
        if self.evidence.candidate.label != self.method:
            raise ValueError("method evidence is not bound to the request")
        return self


class LintClaim(_StrictModel):
    claim_id: str = Field(pattern=r"^lint-[0-9a-f]{16}$")
    quote: str = Field(min_length=1, max_length=600)
    claim_type: Literal["assumption", "analogy", "causal_judgment"]
    structure: str = Field(min_length=1, max_length=800)
    failure_mode: str = Field(min_length=1, max_length=800)
    review_priority: Literal["high", "medium", "low"]
    suggestion: str = Field(min_length=1, max_length=800)
    reference_candidate: Optional[CandidateReference]
    evidence: CandidateEvidenceEnvelope

    @model_validator(mode="after")
    def bind_screen(self) -> "LintClaim":
        if self.evidence.candidate.label != self.quote:
            raise ValueError("lint evidence is not bound to its quote")
        return self


class StructLintResponse(_StrictModel):
    contract_version: Literal["secondary-tools-v2"]
    request_id: str = Field(pattern=r"^[A-Za-z0-9][A-Za-z0-9_-]{11,63}$")
    screening_kind: Literal["internal_ai_document_screen"]
    summary: str = Field(min_length=1, max_length=1_200)
    claims: list[LintClaim] = Field(max_length=30)
    evidence: CandidateEvidenceEnvelope

    @model_validator(mode="after")
    def unique_claims(self) -> "StructLintResponse":
        ids = [claim.claim_id for claim in self.claims]
        if len(ids) != len(set(ids)):
            raise ValueError("claim ids must be unique")
        if self.evidence.candidate.label != "用户提交的策略文档":
            raise ValueError("lint evidence is not bound to the submitted document")
        return self


__all__ = [
    "CONTRACT_VERSION",
    "REQUEST_ID_PATTERN",
    "CandidateEvidenceEnvelope",
    "CandidateReference",
    "StressTestResponse",
    "DiagnoseResponse",
    "MethodApplyResponse",
    "StructLintResponse",
    "ensure_request_id",
    "secondary_scope_guard",
    "internal_screen_evidence",
    "kb_candidate_evidence",
]
